lasagne/layers/shape.py

```diff
@@ -300,3 +302,64 @@ def get_output_for(self, input, **kwargs):
         return padding.pad(input, self.width, self.val, self.batch_ndim)
 
 pad = PadLayer  # shortcut
+
+
+class SliceLayer(Layer):
+    """
+    Slices the input at a specific axis and at specific indices.
+
+    Parameters
+    ----------
+    incoming : a :class:`Layer` instance or a tuple
+        The layer feeding into this layer, or the expected input shape
+
+    indices : int, list, tuple or slice instance
+        When indices is an int the corresponding index is selected from the
+        axis dimension and the axis dimension is dropped. When indices is list,
+        tuple or slice instance the axis dimension is kept even if it is
+        singleton dimension after slicing. Negative values behaves as in numpy
+        indexing.
```

Comment:
I think this misses some bits and explains others in too much detail. What about:
```
If an ``int``, selects a single element from the given axis, dropping the axis. If a slice, selects all elements in the given range, keeping the axis. If a list or a tuple of ``start, stop, step``, will be converted to a slice.
```
Maybe we shouldn't allow tuples/lists because they could be expected to perform fancy indexing instead, especially along with the keyword name `indices`. And the explanation of what happens for "singleton dimensions" is probably best transported in the usage example! It's fully consistent with any indexing in Python, so I wouldn't put special emphasis on that here.